### ISSUE TITLE
Suggestion: Include Sentiment.bin or steps to create it?

### ISSUE THREAD

**Issue by eformx** [knuppe/SharpNL#12]
Hi knuppe:  

1. After an exhaustive search, It looks like there are no OpenNLP sentiment models available for download.  It's surprising really that the hottest thing in NLP is sentiment analysis and no one has a sample bin to play with...I see an opportunity for this library/project.  If you have a en-sentiment.bin file that users could try, just to get them up to speed with their projects that would be fantastic (include it in the models folder).  Having a sentiment model for sandbox environments would really put this library on the map.

2. If we can't find a suitable sentiment model, then what are the steps to create the model?  

The post below (see URL) talks about building a model; but not sure if these build methods are in the SharpNL library?  

3. I suggest that we build an online tool based on SharpNL that has the ability to create sentiment models...based on a users specific needs (upload data, map to tool, go!).  All I need are the steps required to create the training file/model for the DocumentCategorizer. If I can build it on my dev server, I will commit to an online tool for this project.  Just to show what kind of work I do, checkout sankeybuilder.com, a tool I created and got me in Wikipedia as the first web based tool for building Sankey diagrams (ahead of Google, lol). 

Do similar steps exist for SharpNL?
http://stackoverflow.com/questions/21617540/how-to-update-an-existing-named-entity-recognition-model-rather-than-creatin

**Comment by knuppe** [comment]
Hello,

SharpNL supports everything that the current version of OpenNLP has, but unfortunately OpenNLP has no exclusive tool for sentiment analysis. Create the model is not  so complicated, the real problem is to create a tool that does the sentiment analysis :disappointed: 

But there is a light at the end of the tunnel, you can use the `DocumentCategorizer` to categorize a  text into two categories: "good" and "bad" (and use the each category probability). Just keep in mind that this type of sentiment analysis is very superficial.

You can train models using the SharpNL easily (look the method `DocumentCategorizerME.Train`).  I suggest to you to look at the OpenNLP [documentation] (http://opennlp.apache.org/documentation/1.6.0/manual/opennlp.html) and also look the `SharpNL.Tests`. This two references can provide a north to you.


Late last year I started prototyping a tool that improves previously trained models, but I ended up not continuing due to my lack of time to work in the library. I had thought of something a little better than this addon you posted the link :sunglasses: 
Later on I'll continue to develop it, I still have local branch with my prototypes.

**Comment by eformx** [comment]
Hi knuppe,

I created a model from a sampling of tweets where each line is either 0 = negative and 1 = positive.  Tested it and works great (see code below). 

However how do I create the bin file from the model? In the OpenNLP examples there is a serialize method  however I am not sure how to use it? I have tried the BinaryFormatter I don't see that in the Object Browser in Visual Studio. Hope you can shed some light on how to create the bin.

Thanks,

Rob

    Public Shared Function CreateSampleStream() As IObjectStream(Of DocumentSample)
        Dim curDir = HttpContext.Current.Server.MapPath("")
        Dim in_stream As System.IO.Stream = New System.IO.FileStream(curDir & "\tweets.txt", System.IO.FileMode.Open, System.IO.FileAccess.Read)
        Return New DocumentSampleStream(New PlainTextByLineStream(in_stream, "UTF-8"))
    End Function

    Public Shared Function Train(Optional factory As DocumentCategorizerFactory = Nothing) As DocumentCategorizerModel
        Dim model = DocumentCategorizerME.Train("x-unspecified", CreateSampleStream(), TrainingParameters.DefaultParameters(), If(factory, New DocumentCategorizerFactory()))
        Return model
    End Function

     Public Shared Function nlp(ByVal text As String) As String
        nlp = ""
        Dim curDir As String = HttpContext.Current.Server.MapPath("")
       'train DocumentCategorizer Model!
        Dim model As DocumentCategorizerModel = Train()

        'test the model
        Dim classificationME As New DocumentCategorizerME(model)
        Dim modelEvaluator As New DocumentCategorizerEvaluator(classificationME)
        Dim expectedDocumentCategory As String = "Movies"
        Dim documentContent As String = text
        Dim sample As New DocumentSample(expectedDocumentCategory, documentContent)
        Dim classDistribution() As Double = classificationME.Categorize(documentContent)
        Dim predictedCategory As String = classificationME.GetBestCategory(classDistribution)

        'output the result, all works great!
        Dim result As String = documentContent & " " & " Model prediction : " & predictedCategory & "     (1=pos,0=neg)"

       'train DocumentCategorizer Model!
        Dim model As DocumentCategorizerModel = Train()

        'test the model
        Dim classificationME As New DocumentCategorizerME(model)
        Dim modelEvaluator As New DocumentCategorizerEvaluator(classificationME)
        Dim expectedDocumentCategory As String = "Movies"
        Dim documentContent As String = text
        Dim sample As New DocumentSample(expectedDocumentCategory, documentContent)
        Dim classDistribution() As Double = classificationME.Categorize(documentContent)
        Dim predictedCategory As String = classificationME.GetBestCategory(classDistribution)

        'output the result, all works great!
        Dim result As String = documentContent & " " & " Model prediction : " & predictedCategory & " (1=pos,0=neg)"

    'save model to bin
        'Create the stream to add the object(model)
        Dim ms As System.IO.Stream = File.OpenWrite(curDir & "\en-sentiment.bin")
        'Format the object as Binary

        Dim formatter As New BinaryFormatter()
        'serialize object
        formatter.Serialize(ms, model)
        ms.Flush()
        ms.Close()
        ms.Dispose()

However when I try to serialize the model I get the following exception.  Hope you can provide an example of how to save the model to a bin file?

Thanks,

Rob

    'System.Runtime.Serialization.SerializationException was unhandled by user code
        '        HResult = -2146233076
        '        Message = Type 'SharpNL.DocumentCategorizer.DocumentCategorizerModel' in Assembly 'SharpNL, Version=1.1.0.0, Culture=neutral, PublicKeyToken=null' is not marked as serializable.
        '        Source = mscorlib
        'StackTrace:
        'at System.Runtime.Serialization.FormatterServices.InternalGetSerializableMembers(RuntimeType type)
        'at System.Runtime.Serialization.FormatterServices.GetSerializableMembers(Type type, StreamingContext context)
        '        at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitMemberInfo()
        'at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitSerialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
        'at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.Serialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
        'at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Serialize(Object graph, Header[] inHeaders, __BinaryWriter serWriter, Boolean fCheck)
        'at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(Stream serializationStream, Object graph, Header[] headers, Boolean fCheck)
        'at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(Stream serializationStream, Object graph)
        'at _Default.nlp(String text) in C:\Sites\WebSite1\Default.aspx.vb:line 198

**Comment by knuppe** [comment]
This should work.
```vb
Using file As New FileStream(curDir & "\en-sentiment.bin", FileMode.Create)
    model.Serialize(file)
End Using
```

**Quick warning:** The training process is very heavy (it uses a lot of CPU and memory), two or three users using the training process can clog your whole web server.

You should train your models localy and then use in your website. Even so, the process will be quite expensive to the machine.

**Comment by eformx** [comment]
Hi knuppe,

didn't work.  It's because the serialize method does not exist in SharpNL library for the returned object in 'model' (which is DocumentCategorizerModel).  I checked with the docatmodel in the java version and they have a serialize method...

https://opennlp.apache.org/documentation/1.5.3/apidocs/opennlp-tools/opennlp/tools/doccat/DoccatModel.html

Is there another way to serialize the model? or is a fix required to the library?

Thanks for any help on this.

**Comment by eformx** [comment]
If I try it in C# with the following code (some from SharpNL Tests)  it does not work?  Can you try it?  I get the same thing in VB.Net, Serialize method not available...

![image](https://cloud.githubusercontent.com/assets/1794620/8890830/2f87ec42-32dd-11e5-85b7-73395a0fe8be.png)

using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;

using System.IO;
using SharpNL.DocumentCategorizer;
using SharpNL.Tokenize;
using SharpNL.Utility;


public partial class Default2 : System.Web.UI.Page
{


    protected void Page_Load(object sender, EventArgs e)
    {

        IObjectStream<DocumentSample> samples = new GenericObjectStream<DocumentSample>(new[] {
                new DocumentSample("1", new[] {"a", "b", "c"}),
                new DocumentSample("1", new[] {"a", "b", "c", "1", "2"}),
                new DocumentSample("1", new[] {"a", "b", "c", "3", "4"}),
                new DocumentSample("0", new[] {"x", "y", "z"}),
                new DocumentSample("0", new[] {"x", "y", "z", "5", "6"}),
                new DocumentSample("0", new[] {"x", "y", "z", "7", "8"})
            });

        var param = new TrainingParameters();
        param.Set(Parameters.Iterations, "100");
        param.Set(Parameters.Cutoff, "0");

        var model = DocumentCategorizerME.Train("x-unspecified", samples, param, new DocumentCategorizerFactory());

        var doccat = new DocumentCategorizerME(model);

        var aProbs = doccat.Categorize("a");



        dynamic curDir = HttpContext.Current.Server.MapPath("");

        using (FileStream myfile = new FileStream(curDir + "\\en-sentiment.bin", FileMode.Create))
        {
            model.Serialize(myfile);
        }

    }

    public static IObjectStream<DocumentSample> CreateSampleStream()
    {
        dynamic curDir = HttpContext.Current.Server.MapPath("");
        System.IO.Stream in_stream = new System.IO.FileStream(curDir + "\\tweets.txt", System.IO.FileMode.Open, System.IO.FileAccess.Read);
        return new DocumentSampleStream(new PlainTextByLineStream(in_stream, "UTF-8"));
    }

    public static DocumentCategorizerModel Train(DocumentCategorizerFactory factory = null)
    {
        return DocumentCategorizerME.Train("en", CreateSampleStream(), TrainingParameters.DefaultParameters, );
    }
    }

**Comment by knuppe** [comment]
I released the SharpNL 1.1.3 via NuGet that has the `Serialize` method exposed (sorry, I did not see that the method was as internal... my bad! )

Your code is ok for testing purposes.

**Comment by eformx** [comment]
Thanks knuppe!  That worked!  I have a pretty high end developer box with an Intel i7 quad core and 6db ram.  I have another production server with 90gb ram (Windows Server 2012)...I am going to build a bunch of models from dbpedia data and will let you know how long they take so users haves an idea. Shall keep you posted.
